homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,542 @@
+                      discovery_info.get('ssdp_description'),
+                      discovery_info['upnp_device_type'])
+        return
+
+    if config.get(CONF_URL) is not None:
+        url = config.get(CONF_URL)
+        name = config.get(CONF_NAME)
+    elif discovery_info is not None:
+        url = discovery_info['ssdp_description']
+        name = discovery_info['name']
+
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+
+    requester = build_requester(hass)
+
+    # ensure view has been started
+    with (await SETUP_LOCK):
+        # discovered components don't get default values in config
+        config[CONF_LISTEN_PORT] = config.get(CONF_LISTEN_PORT,
+                                              DEFAULT_LISTEN_PORT)
+        server_host, server_port = determine_listen_ip(url, config)
+        notify_view = await async_start_notify_view(hass,
+                                                    server_host,
+                                                    server_port,
+                                                    requester)
+
+    # create device
+    from async_upnp_client import UpnpFactory
+    from async_upnp_client.dlna import DmrDevice
+    factory = UpnpFactory(requester, disable_state_variable_validation=True)
+    try:
+        upnp_device = await factory.async_create_device(url)
+    except (asyncio.TimeoutError, aiohttp.ClientError):
+        raise PlatformNotReady()
+
+    dlna_device = DmrDevice(upnp_device, notify_view.event_handler)
+    device = DlnaDmrDevice(hass, dlna_device, name)
+
+    _LOGGER.debug("Adding device: %s", device)
+    async_add_devices([device])
+
+
+class UpnpNotifyView(HomeAssistantView):
+    """Callback view for UPnP NOTIFY messages."""
+
+    url = '/api/dlna_dmr.notify'
+    name = 'api:dlna_dmr:notify'
+    requires_auth = False
+
+    def __init__(self, base_url):
+        """Initializer."""
+        self.base_url = base_url
+        self.event_handler = None
+        self._registered_services = {}
+        self._backlog = {}
+
+    def register(self, app, router):
+        """Register the view with a router."""
+        handler = request_handler_factory(self, self.async_notify)
+        router.add_route('notify', UpnpNotifyView.url, handler)
+
+    async def async_notify(self, request):
+        """Callback method for NOTIFY requests."""
+        headers = request.headers
+        body = await request.text()
+        status = await self.event_handler.handle_notify(headers, body)
+        return aiohttp.web.Response(status=status)
+
+    @property
+    def callback_url(self):
+        """Full URL to be called by device/service."""
+        return urllib.parse.urljoin(self.base_url, self.url)
+
+
+def build_requester(hass):
+    """Build a derived instance of UpnpRequester, specific for hass."""
+    from async_upnp_client import UpnpRequester
+
+    class HassUpnpRequester(UpnpRequester):
```

Comment:
Why do *we* need to define this class?

The requester instance is only used by the interface library. I would expect that library to define this class. The class needs to except the aiohttp session from hass and we just instantiate the class in this module.

Reply:
To provide de-coupling from libraries. Now, another developer using async_upnp_client can choose to use requests instead of aiohttp, or another future library.

But I'll extend async_upnp_client a bit to provide it as a utility.